Optional Position-Tracking Zipper

xsc opened the pull request:
This PR adds a switch for the position-tracking zipper implementation (defaulting to `false`), as well as the following functions:

- `with-positional-zipper` to use the positional zipper for every call to rewrite-clj's zipper functions,
- `without-positional-zipper` to force usage of `clojure.zip` for every call to rewrite-clj's zipper functions,
- `use-positional-zipper!` to globally enable the positional zipper.

This will reduce the impact on existing applications that mix and match `rewrite-clj.zip` and `clojure.zip` while still providing a reliable implementation of the `position` function when desired.

xsc commented:
This PR would also remove the last concern I have about releasing `0.5.0`. Of course, it's a bit more tedious to use the positional functions but IMO backwards-compatibility and avoiding surprises should have preference.

What do you think?

eraserhd commented:
Sorry about the delay... but...

Ah, neat!

I think one problem with the dynamic variable could be if two different people clients expect two different interfaces, but I think this could be minimized...

**Most** cases should be able to switch implementation using `vector?` or `map?` (or even a keyword from the map).  Then the operation of the zipper would be consistent, and the dynamic variable would only be used for construction.

At that point (this part is a bit nit-picky), we could have either a separate set of constructors, or an extra option key in the constructors, to track position, and remove the dynamic variable altogether.

Them's my thoughts, for what they're worth.

xsc commented:
Thanks for the feedback! I think you're correct in that it should be possible to get rid of the dynamic variable. Let me see what I can cook up. :)

xsc commented:
@eraserhd I added `::custom?` to the custom zipper map and rewrote the zipper fns to dispatch on that value. No dynamic variables, no `with-*`/`without-*` macros.

eraserhd commented:
@xsc works for me! :shipit:
